lib/task_listener.js

```diff
@@ -13,166 +13,275 @@ var request = require('superagent-promise');
 
 /**
 @param {Configuration} config for worker.
 */
-function TaskListener(runtime) {
-  this.runtime = runtime;
-  EventEmitter.call(this);
-}
+export default class TaskListener extends EventEmitter {
+  constructor(runtime) {
+    this.pending = 0;
+    this.runtime = runtime;
+    this.capacity = runtime.capacity;
+    this.runningTasks = [];
+    this.taskQueue = new TaskQueue(this.runtime);
+    this.taskPollInterval = this.runtime.taskQueue.pollInterval;
+    this.claimConfig = {
+      workerId: this.runtime.workerId,
+      workerGroup: this.runtime.workerGroup
+    };
+    super();
+  }
 
-TaskListener.prototype = {
-  __proto__: EventEmitter.prototype,
+  listenForShutdowns() {
+    // If node will be shutdown, stop consuming events.
+    if (this.runtime.shutdownManager) {
+      this.runtime.shutdownManager.once(
+        'nodeTermination', () => {
+          debug('nodeterm');
+          async () => {
+            await this.pause();
+            for(let task of this.runningTasks) {
+              task.abort('worker-shutdown');
+            }
+          }();
+        }.bind(this)
+      );
+    }
+  }
 
-  /**
-  Number of running tasks...
-  */
-  pending: 0,
+  async cancelTask(message) {
+    var runId = message.payload.runId;
+    var reason = message.payload.status.runs[runId].reasonResolved;
+    if (reason !== 'canceled') return;
 
-  connect: function* () {
-    this.runtime.log('listener connect');
+    var taskId = message.payload.status.taskId;
+    var task = this.runningTasks.find(
+      (task) => { return (task.status.taskId === taskId && task.runId === runId); }
+    );
 
-    var self = this;
-    var queue = this.runtime.queue;
+    if (!task) debug('task not found to cancel'); return;
 
-    // Share the queue between all workerTypes of the same provisioner.
-    var queueName;
-    if (this.runtime.createQueue) {
-      queueName =
-        QUEUE_PREFIX +
-        this.runtime.provisionerId + '/' + this.runtime.workerType;
-    }
+    this.runtime.log('cancelling task', {taskId: message.payload.status.taskId});
+    task.cancel(reason);
+  }
+
+  async listenForCancelEvents() {
+    var queue = this.runtime.queue;
 
     var queueEvents = new taskcluster.QueueEvents();
 
-    // Build the listener.
-    var listener = this.listener = new taskcluster.PulseListener({
-      prefetch:     this.runtime.capacity,
-      credentials:  this.runtime.pulse,
-      // Share the queue between all provisonerId + workerTypes.
-      queueName:    queueName
-      // TOOD: Consider adding maxLength.
+    var cancelListener = new taskcluster.PulseListener({
+      credentials: this.runtime.pulse
     });
 
-    yield listener.bind(queueEvents.taskPending({
+    await cancelListener.bind(queueEvents.taskException({
+      workerId: this.runtime.workerId,
       workerType: this.runtime.workerType,
+      workerGroup: this.runtime.workerGroup,
       provisionerId: this.runtime.provisionerId
     }));
```

Comment:
We shouldn't crash on failure to keep the connection to pulse open. As we're really _only_ using it for cancel messages (which is best effort), we should listen for `cancelListener.on('error', (err) => {... });`.

Ideally we could restart the connection and things, but it's a lot of work and it seems better to do this at `PulseListener` level. I want to add support for it there I just haven't figured out how to do so yet; it should probably be hidden behind a option.

Anyways, IMO we should just log the error and then ignore it.. Accepting that for the rest of this docker-worker lifecycle we're not listening for taskExceptions anymore. Then we can add the reconnect option when we implement that in `PulseListener`.